haskell.nix/flake.nix

```diff
@@ -0,0 +1,116 @@
+    haskell-nix.url = "github:serokell/haskell.nix/serokell-latest";
+    hackage-nix = {
+      url = "github:input-output-hk/hackage.nix";
+      flake = false;
+    };
+    stackage-nix = {
+      url = "github:input-output-hk/stackage.nix";
+      flake = false;
+    };
+
+    # if you want to run weeder:
+    ##haskell-nix-weeder = {
+    ##  url = "github:serokell/haskell-nix-weeder";
+    ##  flake = false;
+    ##};
+  };
+
+  outputs = { self, nixpkgs, haskell-nix, hackage-nix, stackage-nix, ... }:
+
+  # for weeder:
+  ##outputs = { self, nixpkgs, haskell-nix, hackage-nix, stackage-nix, haskell-nix-weeder, ... }:
+
+    let
+      haskellNix = import haskell-nix {
+        sourcesOverride = { hackage = hackage-nix; stackage = stackage-nix; };
+      };
+
+      pkgs = import nixpkgs haskellNix.nixpkgsArgs;
+      lib = pkgs.lib;
+
+      # invoke haskell.nix
+      hs-pkgs = pkgs.haskell-nix.stackProject {
+        src = pkgs.haskell-nix.haskellLib.cleanGit {
+          name = "pataq-package";
+          src = ./.;
+        };
+
+        # haskell.nix configuration
+        modules = [{
+          packages.pataq-package = {
+            ghcOptions = [
+              # fail on warnings
+              "-Werror"
+              # disable optimisations, we don't need them if we don't package or deploy the executable
+              "-O0"
+
+              # for weeder: produce *.dump-hi files
+              ##"-ddump-to-file" "-ddump-hi"
+            ];
+
+            # for weeder: collect all *.dump-hi files
+            ##postInstall = weeder-hacks.collect-dump-hi-files;
+          };
+
+        }];
+      };
+
+      hs-pkg = hs-pkgs.pataq-package;
+
+      # returns the list of all components for a package
+      get-package-components = pkg:
+        # library
+        lib.optional (pkg ? library) pkg.library
+        # haddock
+        ++ lib.optional (pkg ? library) pkg.library.haddock
+        # exes, tests and benchmarks
+        ++ lib.attrValues pkg.exes
+        ++ lib.attrValues pkg.tests
+        ++ lib.attrValues pkg.benchmarks;
+
+      # all components for the current haskell package
+      all-components = get-package-components hs-pkg.components;
+
+      # for weeder:
+      ##weeder-hacks = import haskell-nix-weeder { inherit pkgs; };
+
+      # nixpkgs has weeder 2, but we use weeder 1
+      ##weeder-legacy = pkgs.haskellPackages.callHackageDirect {
+      ##  pkg = "weeder";
+      ##  ver = "1.0.9";
+      ##  sha256 = "0gfvhw7n8g2274k74g8gnv1y19alr1yig618capiyaix6i9wnmpa";
+      ##} {};
+
+      # a derivation which generates a script for running weeder
+      ##weeder-script = weeder-hacks.weeder-script {
+      ##  weeder = weeder-legacy;
+      ##  hs-pkgs = hs-pkgs;
+      ##  local-packages = [
+      ##    { name = "pataq-package"; subdirectory = "."; }
+      ##  ];
+      ##};
+
+    in {
+      # nixpkgs revision pinned by this flake
+      legacyPackages.x86_64-linux = pkgs;
```

Comment:
Maybe consider `flake-utils.lib.eachSystem [ "x86_64-linux" ]` instead of hardcoding, so that if we need to also do this for `x86_64-darwin` we can

Reply:
I think it's better to keep the template simpler and not add `eachSystem` if we don't use it, it can be especially problematic when you want to add a flake output that does not depend on the system

Reply:
Well, we do have Haskell projects which must have `x86_64-darwin` outputs (e.g. bridge-web), so if we want to use this template there it will become a problem.

> it can be especially problematic when you want to add a flake output that does not depend on the system

Perhaps add an empty set in which non-system-dependent outputs go? Like
```
outputs = {...}:
  {
    # put non-system-dependent outputs here
  } // flake-utils.lib.eachSystem ....
```

Reply:
The problem is that with `eachSystem` the common layout is to use `let` like
```
outputs = {...}:
  {
    ...
  } // flake-utils.lib.eachSystem (system: let ... in { ... })
```

And when you want to add non-system-dependent outputs, you often want to use definitions from the `let` block, but they are not available because they are defined with dependency on `system`. And it can get confusing to refactor, to make the code look sensible. So I would avoid `eachSystem` in the template, and if we actually need it in some projects, we can add it there.

Also, in some cases, I'd prefer to define outputs for other systems explicitly, e.g. define `packages.x86_64-linux` and `packages.x86_64-darwin`, it should be simpler to grasp for people new to nix.

Reply:
I would argue that the difficulty of this is yet another good reason to iterate systems, by not doing so we're shipping an incomplete template which people will write code for which will become an awkward refactoring task once they eventually need to support different systems, if they are working with it from the beginning correctly then this will be less of an issue.

Reply:
There are a lot of other reasons people might need to refactor this template: using a stack project with multiple packages, adding non-haskell packages, adding extra options for the build, adding extra checks to the CI. So I'd keep this template as simple as possible, and deal with the situations where it's not sufficient separately, because trying to cover more usecases will make it harder to understand, and will make it harder to make changes which conflict wtih a cover-all template, e.g. when they want to make changes to the build which are not compatible with a specific system.

Even for situations where people do want to add definitions for non-linux systems, `eachSystems` may not be the best choice sometimes, like if you only want to maintain one or two outputs for non-linux systems, but most of the outputs you only need for linux, then it's better to add an output like `packages.x86_64-darwin` explicitly.

Reply:
I still don't think that makes sense to me, even if you want to list platforms explicitly you will still want to reuse a lot of the code, so you would still need it to be wrapped in a function which consumes `system` regardless of how you want to implement producing the attribute set, and conveniently `flake-utils` supports that too as @balsoft initially mentioned.

I also don't think supporting extra platforms suffices as an edge case enough to dismiss it along with other things that will require a refactor, most other things you mentioned are dependent on the way the project is made, whereas most projects are by default in theory compatible with more than one platform, so it only makes sense that supporting a new platform explicitly should only involve a single line of code modification.

It is also something that may not be immediately obvious to developers if not provided by example, so including it is important to discourage copy pasting large chunks of Nix.

Reply:
> even if you want to list platforms explicitly you will still want to reuse a lot of the code, so you would still need it to be wrapped in a function which consumes system regardless of how you want to implement producing the attribute set

You don't necessarily have to wrap `outputs` into a function consuming `system` to achieve that, you can also make definitions in the `let` block accept `system` argument, and then make separate outputs for different systems which would call that with a different `system`. This approach might be preferable when you have a lot of outputs, but there are only one or two of them which you want to maintain for different systems.

> I also don't think supporting extra platforms suffices as an edge case enough to dismiss it along with other things that will require a refactor

What I'd say is that it's a usecase we rarely run into, in ~25 haskell.nix projects we've made so far, I can only remember ligo, morley-debugger, and I believe something in stakerdao, where we run haskell.nix builds on non-linux systems. And for example, in both ligo and morley-debugger it wouldn't be enough to just add "x86_64-darwin" to the list of systems, because they both use it for packaging a vscode extension, which requires a lot of extra code by itself.

> most projects are by default in theory compatible with more than one platform, so it only makes sense that supporting a new platform explicitly should only involve a single line of code modification

"Single line of code" is far from true in practice, we often had situations where adding support for a new platform would be non-trivial and required digging into haskell.nix or coming up with some workarounds. We can't rely on haskell.nix to just work when we add a new system. We could also have to use some system-specific tools in the definitions ourselves, e.g. packages which are hard to get for non-linux systems.

> It is also something that may not be immediately obvious to developers if not provided by example, so including it is important to discourage copy pasting large chunks of Nix.

Still there are also many other features we often use in CI, it would need a lot of work to make a template to cover all of them. And some of them may just conflict with each other and require a lot of work to be supported at the same time. So I'd start with a template which covers just the most basic ones, and discuss everything else after we've done with that.